programs/x509/cert_write.c

```diff
@@ -803,6 +842,357 @@ int main( int argc, char *argv[] )
+ *  Receives:
+ *      <4 bytes length indicator in network order><concatenated public key>
+ */
+int load_pubkey_from_remote( const char * remote_info, mbedtls_pk_context * ctx )
+{
+    int key_idx = 0, offset = 0, ret = 0;
+    const char * serial_port = NULL;
+    unsigned char func_buffer[10];
+    unsigned char pub_key_buf[100];
+    size_t rx_len = 0;
+    static mbedtls_ecp_keypair ecp_key;
+
+    offset = strlen( REMOTE_KEY_CMD_TAG );
+    key_idx = (int)remote_info[offset++];
+    key_idx = key_idx - 48; // ascii to decimal
+
+    if ( key_idx < REMOTE_KEY_ID_MIN || key_idx > REMOTE_KEY_ID_MAX )
+    {
+        mbedtls_printf( " failed\n  !  Invalid remote key index %d\n\n", key_idx );
+        return( -1 );
+    }
+    serial_port = remote_info + offset;
+
+    /* Prepare command */
+    offset = 0;
+    func_buffer[offset++] = REMOTE_KEY_FUNC_GET_PUBKEY;
+    func_buffer[offset++] = key_idx;
+
+    if ( serial_xfer( serial_port, func_buffer, offset, pub_key_buf, sizeof( pub_key_buf ), &rx_len ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  Serial error trying to get pulic key\n\n" );
+        return( -1 );
+    }
+
+    /* Import public key from received binary */
+    mbedtls_ecp_keypair_init(&ecp_key);
+    ret = mbedtls_ecp_group_load(&ecp_key.grp, MBEDTLS_ECP_DP_SECP256R1);
+    if ( ret != 0 )
+        return( -1 );
+    ret = mbedtls_ecp_point_read_binary(&ecp_key.grp, &ecp_key.Q, pub_key_buf, rx_len );
+    if ( ret != 0 )
+    {
+        mbedtls_printf( " failed\n  !  Failed to read ecp key from binary\n\n" );
+        return( -1 );
+    }
+    ctx->pk_info = &mbedtls_eckey_info;
+    ctx->pk_ctx = &ecp_key;
+    return( 0 );
+}
+
+/**
+ * @brief           Tell if can do the operation given by type
+ *
+ * @param type      Target type
+ *
+ * @return          0 if context can't do the operations,
+ *                  1 otherwise.
+ */
+static int remote_can_do_func(const void *ctx, mbedtls_pk_type_t type)
+{
+    UNUSED(ctx);
+    /* At the moment on ECDSA is supported */
+    return (MBEDTLS_PK_ECDSA == type);
+}
+
+/**
+  * @brief  Use STSAFE private key for signature.
+  *
+  * @param ctx       ECDSA context
+  * @param md_alg    Algorithm that was used to hash the message
+  * @param hash      Message hash
+  * @param hash_len  Length of hash
+  * @param sig       Buffer that will hold the signature
+  * @param sig_len   Length of the signature written
+  * @param f_rng     RNG function
+  * @param p_rng     RNG parameter
+  *
+  * @retval 0 if successful, or 1.
+  */
+static int remote_sign_func(void *ctx, mbedtls_md_type_t md_alg,
+                            const unsigned char *hash, size_t hash_len,
+                            unsigned char *sig, size_t *sig_len,
+                            int (*f_rng)(void *, unsigned char *, size_t),
+                            void *p_rng)
+{
+    remote_serial_pk_context * remote_ctx = (remote_serial_pk_context *)ctx;
+    unsigned char func_buffer[1024];
```

Comment:
Please use `MBEDTLS_MD_MAX_SIZE + (whatever the overhead is)` rather than a magic constant.

Or use 32 bytes + overhead, with a note explaining why, since this function is currently limited to SHA-256.